### ISSUE TITLE
A runtime error occured on assigning an enum to an interface:  System.NullReferenceException

### ISSUE THREAD

**Issue by DBiernat** [microsoft/AL#5926]
**Describe the bug**
Assigning an enum to an interface ends up in a runtime error, if the enum field is a flowfield and the source and target fields have different types. If they have the same type it work's as expected.

```
The description for Event ID 0 from source MicrosoftDynamicsNAVClientWebClient cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

If the event originated on another computer, the display information had to be saved with the event.

The following information was included with the event: 

Error accessing Website 
Type: Microsoft.Dynamics.Nav.Types.Exceptions.NavServerSessionException
Message: The server "BC" was unable to process the request. The application will close.
StackTrace:
   at Microsoft.Dynamics.Nav.Client.ConnectionStateManager.CallServer[T](BeginCallServerMethod beginCallServerMethod, EndCallServerMethod`1 endCallServerMethod, ServerCallOptions options) in s:\repo\src\Platform\Client\Shared\Prod.Client.ServiceConnection\ConnectionStateManager.cs:line 145
   at Microsoft.Dynamics.Nav.Client.ServiceConnection.CallServer[T](BeginCallServerMethod beginCallServerMethod, EndCallServerMethod`1 endCallServerMethod) in s:\repo\src\Platform\Client\Shared\Prod.Client.ServiceConnection\ServiceConnection.cs:line 2120
   at Microsoft.Dynamics.Nav.Client.ServiceConnectionBase.ValidateField(NavRecordState& state, NavDataSet recDataSet, Int32 controlId) in s:\repo\src\Platform\ClientServerShared\Prod.Types\ServiceConnection\ServiceConnectionBase.cs:line 446
   at Microsoft.Dynamics.Nav.Client.DataBinder.NstDataAccess.ValidateField(ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Databinder\NSTDataAccess.cs:line 597
   at Microsoft.Dynamics.Nav.Client.DataBinder.NavBindingManager.ValidateField(ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Databinder\NavBindingManager.cs:line 1506
   at Microsoft.Dynamics.Nav.Client.Validation.RowEntryValidateItem.ValidateItem(ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Validation\RowEntryValidateItem.cs:line 44
   at Microsoft.Dynamics.Nav.Client.Validation.ValidationHelper.ExecuteValidation(ValidationItem validationItem, String message) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Validation\ValidationHelper.cs:line 414
   at Microsoft.Dynamics.Nav.Client.Validation.ValidationHelper.ValidateItem(ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Validation\ValidationHelper.cs:line 386
   at Microsoft.Dynamics.Nav.Client.Validation.ValidationHelper.Validate(IValidationCaller validationCaller, ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Validation\ValidationHelper.cs:line 215
   at Microsoft.Dynamics.Nav.Client.DataBinder.NavBindingManager.Validate(IValidationCaller validationCaller, ValidationItem validationItem) in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Databinder\NavBindingManager.cs:line 1444
   at Microsoft.Dynamics.Nav.Client.Validation.FieldValidationRule.<ValidateValue>d__4.MoveNext() in s:\repo\src\Platform\Client\Shared\Prod.Client.UI\Validation\FieldValidationRule.cs:line 61
   at Microsoft.Dynamics.Framework.UI.ValidationRule.ValidateRuleCollection(ValidationResult result, IEnumerable`1 rules, Object value, IValidationCaller validationCaller) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Validation\ValidationRule.cs:line 89
   at Microsoft.Dynamics.Framework.UI.EditLogicalControl.ValidateHost(IValidationCaller validationCaller, ValidationResult validationResult) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Forms\EditLogicalControl.cs:line 822
   at Microsoft.Dynamics.Framework.UI.ValidationHolder.ValidateInternal(IValidationCaller validationCaller, ValidationResult validationResult) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Validation\ValidationHolder.cs:line 159
   at Microsoft.Dynamics.Framework.UI.ValidationHolder.PerformValidation(IValidationCaller validationCaller, ValidationResult validationResult) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Validation\ValidationHolder.cs:line 136
   at Microsoft.Dynamics.Framework.UI.ValidationHolder.Validate(ValidationResult validationResult) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Validation\ValidationHolder.cs:line 120
   at Microsoft.Dynamics.Framework.UI.EditLogicalControl.SetObjectValue(Object value) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Forms\EditLogicalControl.cs:line 469
   at Microsoft.Dynamics.Framework.UI.Int32Control.SetObjectValue(Object value) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Forms\NumberControls.cs:line 427
   at Microsoft.Dynamics.Framework.UI.ValueControl`2.set_StringValue(String value) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Forms\ValueControl.cs:line 132
   at Microsoft.Dynamics.Framework.UI.SaveValueInteraction.InvokeCore(SaveValueInteractionInput logicalInteractionInput) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Interactions\LogicalInteractions\SaveValueInteraction.cs:line 96
   at Microsoft.Dynamics.Framework.UI.InteractionInvoker.Invoke[T](String interactionName, Action`1 invokeMethod, T input) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Interactions\InteractionInvoker.cs:line 37
   at Microsoft.Dynamics.Framework.UI.LogicalInteraction`1.Invoke(T logicalInteractionInput) in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Interactions\LogicalInteraction.cs:line 40
   at Microsoft.Dynamics.Framework.UI.InteractionManager.<>c__DisplayClass8_2.<InvokeInteractions>b__4() in s:\repo\src\Platform\Client\Shared\Prod.ClientFwk\Interactions\InteractionManager.cs:line 126
Source: Microsoft.Dynamics.Nav.Client.ServiceConnection----------------------------------
Type: Microsoft.Dynamics.Nav.Types.Exceptions.NavServerSessionException
Message: The server "BC" was unable to process the request. The application will close.
StackTrace:

Source: ----------------------------------
Type: System.NullReferenceException
Message: Object reference not set to an instance of an object.
StackTrace:
   at Microsoft.Dynamics.Nav.Runtime.NCLEnumMetadata.GetImplementationCodeunitId(Int32 ordinalValue, Int32 interfaceIndex)
   at Microsoft.Dynamics.Nav.Runtime.ALCompiler.ToInterface(ITreeObject parentOfResult, NavOption optionValue, Int32 interfaceIndex)
   at Microsoft.Dynamics.Nav.BusinessApplication.Record50101.Value_a45_OnValidate_Scope.OnRun()
   at Microsoft.Dynamics.Nav.Runtime.NavMethodScope.Run()
   at Microsoft.Dynamics.Nav.BusinessApplication.Record50101.Value_a45_OnValidate()
   at Microsoft.Dynamics.Nav.Runtime.NavRecord.ValidateField(FieldTriggerHandler`1 fieldEvent, NCLMetaField metaField, Boolean isUserInput, List`1 beforeHandlers, List`1 afterHandlers)
   at Microsoft.Dynamics.Nav.Runtime.NavRecord.Validate(NCLMetaField metaField, NavValue newValue, NavRecord callerRecord, Boolean isUserInput)
   at Microsoft.Dynamics.Nav.Service.NsDataAccess.Validate(Int32 fieldId)
   at Microsoft.Dynamics.Nav.Service.NsFormDataAccess.<>c__DisplayClass44_0.<Validate>b__0()
   at Microsoft.Dynamics.Nav.Runtime.ApplicationObjectRootScope.AddApplicationObjectRootScope(NavApplicationObjectBase applicationObject, Action action)
   at Microsoft.Dynamics.Nav.Service.NSField.ValidateField(NavSession session)
   at Microsoft.Dynamics.Nav.Service.NSField.Validate(NavSession session)
   at SyncInvokeValidateField(Object , Object[] , Object[] )
   at System.ServiceModel.Dispatcher.SyncMethodInvoker.Invoke(Object instance, Object[] inputs, Object[]& outputs)
   at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.RunInTransactionCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
   at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass28_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
   at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.TransientErrorRetryCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
   at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.<>c__DisplayClass28_1.<Combine>b__1(NSServiceBase serviceInstance, Object[] inputs, Object[]& outputs)
   at Microsoft.Dynamics.Nav.Service.ServiceOperationInvoker.ErrorMappingCombinator(ServiceOperation innerOperation, NSServiceBase serviceInstance, MethodBase syncMethod, Object[] inputs, Object[]& outputs)
Source: Microsoft.Dynamics.Nav.Ncl
```

**To Reproduce**
Steps and to reproduce the behavior:
1. Create a new docker container based on the images described at the bottom
2. Build, deploy and run the attached project
3. Insert a value into the line field Value
4. By inserting the value the following code will be executed and the error occured on: IType := Type;

```
            trigger OnValidate()
            var
                IType: Interface Type;
            begin
                CalcFields(Type);

                IType := Type;
                IType.ValidateValue();
            end;
 ```

**Expected behavior**
No runtime error should occur or (if this should not be possible) a compiler error must be shown.

**Versions:**
 - Docker Image: mcr.microsoft.com/businesscentral/onprem
 - AL Language: 5.0.280447
 - Business Central: W1 16.0 (Platform 16.0.11233.12061 + Application 16.0.11240.12076)

also tested in

 - Docker Image: bcinsider.azurecr.io/bcsandbox-master
 - AL Language: 6.0.277481
 - Business Central: W1 17.0 (Platform 16.0.13337.0 + Application 17.0.13520.0)

[Bug.zip](https://github.com/microsoft/AL/files/4701071/Bug.zip)

**Comment by DBiernat** [comment]
Just to make sure which is the right be behaviour from your point of view?

**Comment by esbenk** [comment]
The NullReferenceException is never the right behavior:) Based on the "different types" part it sounds like it should have been blocked by compiler or at least fail gracefully at runtime. Can you provide a simple repro?

**Comment by DBiernat** [comment]
> 
> 
> The NullReferenceException is never the right behavior:) Based on the "different types" part it sounds like it should have been blocked by compiler or at least fail gracefully at runtime. Can you provide a simple repro?

Look into the attachment (Bug.zip).

**Comment by esbenk** [comment]
Sorry. I missed that.

**Comment by DBiernat** [comment]
> 
> 
> Sorry. I missed that.

Any news?

**Comment by esbenk** [comment]
This is fixed in 16.1, but it is a server-side fix.
